File: 2024/day19.py
from functools import cache

with open("data/input19.txt") as f:
    data = f.read()

@cache
def valid(patterns, design):
    # if design is empty there is a match
    if not design:
        return True


    for pattern in patterns:
        if design.startswith(pattern):
            remaining = design[len(pattern):]
            if valid(patterns, remaining):
                return True

    # if no match, return false
    return False

def part1(data):
    patterns, designs = data.strip().split("\n\n")
    patterns = tuple(p.strip() for p in patterns.split(','))
    sum = 0

    for design in designs.splitlines():
        if valid(patterns, design):
            sum += 1
    print(sum)

@cache
def count_ways(patterns, design):
    # if empty, all patterns checked and one valid way
    if not design:
        return 1

    total = 0
    for pattern in patterns:
        if design.startswith(pattern):
            remaining = design[len(pattern):]
            ways = count_ways(patterns, remaining)
            total += ways

    return total

def part2(data):
    patterns, designs = data.strip().split("\n\n")
    patterns = tuple(p.strip() for p in patterns.split(','))

    total = 0
    for design in designs.splitlines():
        ways = count_ways(patterns, design)
        total += ways
    print(total)

part1(data)
part2(data)

File: 2024/day19.go
package main

import (
	"fmt"
	"os"
	"strings"
)

type memoKey struct {
	design string
}

func valid(patterns []string, design string, memo map[memoKey]bool) bool {
	key := memoKey{design}

	if result, exists := memo[key]; exists {
		return result
	}

	if len(design) == 0 {
		return true
	}

	for _, pattern := range patterns {
		if strings.HasPrefix(design, pattern) {
			remaining := design[len(pattern):]
			if valid(patterns, remaining, memo) {
				memo[key] = true
				return true
			}
		}
	}

	memo[key] = false
	return false
}

func countWays(patterns []string, design string, memo map[memoKey]int) int {
	key := memoKey{design}

	if count, exists := memo[key]; exists {
		return count
	}

	if len(design) == 0 {
		return 1
	}

	total := 0
	for _, pattern := range patterns {
		if strings.HasPrefix(design, pattern) {
			remaining := design[len(pattern):]
			total += countWays(patterns, remaining, memo)
		}
	}

	memo[key] = total
	return total
}

func Day19() {
	data, err := os.ReadFile("data/input19.txt")
	if err != nil {
		panic(err)
	}

	parts := strings.Split(string(data), "\n\n")
	if len(parts) != 2 {
		panic("Invalid input format")
	}

	patternLine := strings.TrimSpace(parts[0])
	patterns := strings.Split(patternLine, ",")
	for i := range patterns {
		patterns[i] = strings.TrimSpace(patterns[i])
	}

	designs := strings.Split(strings.TrimSpace(parts[1]), "\n")

	validMemo := make(map[memoKey]bool)
	validCount := 0
	for _, design := range designs {
		if valid(patterns, design, validMemo) {
			validCount++
		}
	}
	fmt.Println(validCount)

	waysMemo := make(map[memoKey]int)
	totalWays := 0
	for _, design := range designs {
		ways := countWays(patterns, design, waysMemo)
		totalWays += ways
	}
	fmt.Println(totalWays)
}